kubernetes/deploy.sh

```diff
@@ -11,20 +11,80 @@ usage: $0 [ -r REVERSE-CIDR ] [ -i DNS-IP ] [ -d CLUSTER-DOMAIN ] [ -t YAML-TEMP
          then the default is to handle all reverse zones (i.e. in-addr.arpa and ip6.arpa)
     -i : Specify the cluster DNS IP address. If not specificed, the IP address of
          the existing "kube-dns" service is used, if present.
+    -s : Skips the translation of kube-dns configmap to the corresponding CoreDNS Corefile configuration.
+
 USAGE
 exit 0
 }
 
 # Simple Defaults
 CLUSTER_DOMAIN=cluster.local
 YAML_TEMPLATE=`pwd`/coredns.yaml.sed
+YAML_UPDATED=`pwd`/coredns.yaml
+
+
+cp ${YAML_TEMPLATE} ${YAML_UPDATED}
+
+# Translates the kube-dns ConfigMap to equivalent CoreDNS Configuration.
+function translate-kube-dns-configmap {
+    kube-dns-federation-to-coredns
+    kube-dns-upstreamnameserver-to-coredns
+    kube-dns-stubdomains-to-coredns
+}
+
+function kube-dns-federation-to-coredns {
+  local -r federation_data="/tmp/feddata.txt"
+  local no_data=""
+
+  kubectl -n kube-system get configmap kube-dns  -ojsonpath='{.data.federations}' 2> /dev/null | jq .  | tr -d '":,' > ${federation_data}
+  if [ -s ${federation_data} ]; then
+    sed -e '1s/^/federation /' -e 's/^/        /' -e '1i\\' -i ${federation_data}
+    sed -e "s/FEDERATIONS/$(<${federation_data} sed -e 's/[\&/]/\\&/g' -e 's/$/\\n/' | tr -d '\n')/g" -i ${YAML_UPDATED}
+    rm ${federation_data}
+  else
+    sed -i "s/FEDERATIONS/${no_data}/g" ${YAML_UPDATED}
+ fi
+}
+
+function kube-dns-upstreamnameserver-to-coredns {
+  local -r upstreamdatanameserver_data="/tmp/upstreamdata.txt"
+
+  kubectl -n kube-system get configmap kube-dns  -ojsonpath='{.data.upstreamNameservers}' 2> /dev/null | tr -d '[",]' > ${upstreamdatanameserver_data}
+  if [ -s ${upstreamdatanameserver_data} ]; then
+    sed -e "s/UPSTREAMNAMESERVER/$(<${upstreamdatanameserver_data} sed -e 's/[\&/]/\\&/g' -e 's/$/\\n/' | tr -d '\n')/g" -i ${YAML_UPDATED}
```

Comment:
What does `$(<${` ... do? It's hard to google those things (all non-alphanumeric)